Add mrqa

bhavani105 opened the pull request:
# Pull Request      

## What does this PR do?
Adds support for the HF MRQA dataset
Allows using the entire dataset or subset 
Parameter to run use with run_mrc.py
      --dataset_name mrqa \
      --dataset_config_name plain_text \
      --preprocessor primeqa.mrc.processors.preprocessors.mrqa.MRQAPreprocessor 
Additional parameter to use a named subset:
      --dataset_filter_column_value <subset-name>
      --dataset_filter_column_name subset

Required for https://github.com/primeqa/primeqa/issues/304

  

Notes:
- Replace `(issue)` **above** ↑↑↑ with the issue this PR closes to automatically link the two.
This must be done when the PR is created.
- Add multiple `Closes #(issue)` as needed.
- If this PR is work towards but does not close an issue, simply tag the issue without mentioning `Closes`.

## Description

**Describe** the changes proposed by this PR below to give the reviewer context below ↓↓↓

```(description)```     

### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the (development guide documentation)[development.html].

 - [ ] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

## Releases

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

### Checklist

Review the following and mark as completed:
- [ ] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [ ] Added [description](#description) of changes proposed.
- [ ] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [ ] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

ma-sultan commented:
@bhavani105
1. HotpotQA seems to be missing in the tester script (e.g., in lines 14, 32, 33 and so on).
2. if the preprocessor is able to handle a subset of the MRQA datasets (as opposed to a single one or all of them, e.g., just SQuAD and NewsQA), please provide an example usage of it in the README.
3. It might be worth adding a short description of what happens under the hood when the user provides multiple datasets, e.g., they are first concatenated in the order they appear in the input and then randomly shuffled, thus a training minibatch can generally contain examples from multiple datasets.

bhavani105 commented:
> 

I did not exhaustively list the subsets in MRQA.
- Added a HotpotQA examples to the test case
- Added multiple subsets selection test case
- Updated README.md with multiple datasets selection example 
- Renamed arg to `dataset_filter_column_values`
- Training examples are shuffled.  eval examples are kept in the original order

ma-sultan commented:
LGTM, approving.

ma-sultan reviewed (approved):
LGTM

mihaela-bornea commented on primeqa/ir/dense/colbert_top/colbert/modeling/hf_colbert_xlmr.py:
```diff
@@ -1,10 +1,9 @@
 import torch.nn as nn
```
Do we need both hf_colbert.py and hf_colbert_xlmr.py ?

mihaela-bornea commented on notebooks/ir/dense/dense_ir.ipynb:
```diff
@@ -245,15 +245,34 @@
       "    \"queries\": null,\n",
       "    \"index_name\": null,\n",
```
There might be something wrong with this file because I can not do view file it in github. I get a message saying this is not a valid notebook.

bhavani105 replied:
There was a problem with the rebase - fixed now

mihaela-bornea commented on primeqa/mrc/run_mrc.py:
```diff
@@ -128,6 +129,16 @@ class DataTrainingArguments:
             "help": "The configuration name of the dataset to use (via the datasets library)."
         }
     )
+    dataset_filter_column_name: str = field(
```
This argument needs more documentation and it is not easy to understand what it does. Is it the right place for this argument? Can it be handled by pre-processors ?

mihaela-bornea reviewed (commented):
I left a few questions/comments.

rongzhangibm reviewed (approved):
LGTM.
1. In run_mrc.py, it's better to add a check to make sure both 'dataset_filter_colume_values' and 'dataset_filter_colume_name' show up in the argument or not.
2. Do we have the training/eval results for mrqa?

rongzhangibm commented on primeqa/mrc/run_mrc.py:
```diff
@@ -425,6 +441,10 @@ def main():
+        if data_args.dataset_filter_column_values is not None:
+            logger.info(f"Filter EVAL dataset {data_args.dataset_filter_column_name} {data_args.dataset_filter_column_values}")
+            eval_examples = eval_examples.filter(lambda example: example[data_args.dataset_filter_column_name] in (data_args.dataset_filter_column_values))
+            logger.info(f"Filtered EVAL dataset size {train_dataset.num_rows}")
```
logger.info(f"Filtered EVAL dataset size {eval_examples.num_rows}")

bhavani105 replied:
fixed

rongzhangibm commented on primeqa/mrc/README.md:
````diff
@@ -232,6 +232,22 @@ R@P=0.9: 10.16% (actual p=90.00%, score threshold=7.425)
+ ```shell
+       --dataset_name mrqa \
+       --dataset_config_name plain_text \
+       --preprocessor primeqa.mrc.processors.preprocessors.mrqa.MRQAPreprocessor 
````
I think we also need to add postprocessor and eval_mtrics here to use squad format.

rongzhangibm replied:
Seems that the subset name for TriviaQA  is TriviaQA-web.

bhavani105 replied:
fixed

mihaela-bornea reviewed (approved):
LGTM
